Python/specialize.c

```diff
@@ -2663,101 +2663,102 @@ _Py_Specialize_Send(_PyStackRef receiver_st, _Py_CODEUNIT *instr)
+{
+    PyNumberMethods *nb = ty->tp_as_number;
+    if (nb && nb->nb_bool) {
+        return SPEC_FAIL_TO_BOOL_NUMBER;
+    }
+    PyMappingMethods *mp = ty->tp_as_mapping;
+    if (mp && mp->mp_length) {
+        return SPEC_FAIL_TO_BOOL_MAPPING;
+    }
+    PySequenceMethods *sq = ty->tp_as_sequence;
+    if (sq && sq->sq_length) {
+      return SPEC_FAIL_TO_BOOL_SEQUENCE;
+    }
+    return 0;
+}
+
+
 void
 _Py_Specialize_ToBool(_PyStackRef value_o, _Py_CODEUNIT *instr)
 {
-    assert(ENABLE_SPECIALIZATION);
+    assert(ENABLE_SPECIALIZATION_FT);
     assert(_PyOpcode_Caches[TO_BOOL] == INLINE_CACHE_ENTRIES_TO_BOOL);
     _PyToBoolCache *cache = (_PyToBoolCache *)(instr + 1);
     PyObject *value = PyStackRef_AsPyObjectBorrow(value_o);
     if (PyBool_Check(value)) {
-        instr->op.code = TO_BOOL_BOOL;
-        goto success;
+        specialize(instr, TO_BOOL_BOOL);
+        return;
     }
     if (PyLong_CheckExact(value)) {
-        instr->op.code = TO_BOOL_INT;
-        goto success;
+        specialize(instr, TO_BOOL_INT);
+        return;
     }
     if (PyList_CheckExact(value)) {
-        instr->op.code = TO_BOOL_LIST;
-        goto success;
+        specialize(instr, TO_BOOL_LIST);
+        return;
     }
     if (Py_IsNone(value)) {
-        instr->op.code = TO_BOOL_NONE;
-        goto success;
+        specialize(instr, TO_BOOL_NONE);
+        return;
     }
     if (PyUnicode_CheckExact(value)) {
-        instr->op.code = TO_BOOL_STR;
-        goto success;
+        specialize(instr, TO_BOOL_STR);
+        return;
     }
     if (PyType_HasFeature(Py_TYPE(value), Py_TPFLAGS_HEAPTYPE)) {
-        PyNumberMethods *nb = Py_TYPE(value)->tp_as_number;
-        if (nb && nb->nb_bool) {
-            SPECIALIZATION_FAIL(TO_BOOL, SPEC_FAIL_TO_BOOL_NUMBER);
-            goto failure;
-        }
-        PyMappingMethods *mp = Py_TYPE(value)->tp_as_mapping;
-        if (mp && mp->mp_length) {
-            SPECIALIZATION_FAIL(TO_BOOL, SPEC_FAIL_TO_BOOL_MAPPING);
-            goto failure;
-        }
-        PySequenceMethods *sq = Py_TYPE(value)->tp_as_sequence;
-        if (sq && sq->sq_length) {
-            SPECIALIZATION_FAIL(TO_BOOL, SPEC_FAIL_TO_BOOL_SEQUENCE);
-            goto failure;
+        unsigned int version = 0;
+        int err = _PyType_Validate(Py_TYPE(value), check_type_always_true, &version);
+        if (err < 0) {
+            unspecialize(instr, SPEC_FAIL_OUT_OF_VERSIONS);
+            return;
         }
-        if (!PyUnstable_Type_AssignVersionTag(Py_TYPE(value))) {
-            SPECIALIZATION_FAIL(TO_BOOL, SPEC_FAIL_OUT_OF_VERSIONS);
-            goto failure;
+        else if (err > 0) {
+            unspecialize(instr, err);
+            return;
         }
-        uint32_t version = type_get_version(Py_TYPE(value), TO_BOOL);
+
+        assert(err == 0);
         if (version == 0) {
-            goto failure;
+            unspecialize(instr, SPEC_FAIL_OUT_OF_VERSIONS);
+            return;
         }
```

Comment:
This feels like it might be a bit too defensive, but up to you. By construction,`_PyType_Validate` should never return 0 without setting a version. And if that invariant is violated for some reason then the `assert(version)` below should catch it.